FernFlower-Patches/0024-Add-try-with-resource-support.patch

```diff
@@ -0,0 +1,493 @@
+From 9a1c544d542fcc7d3986b9eaf59bc27aeeb314df Mon Sep 17 00:00:00 2001
+From: Justin <jrd2558@gmail.com>
+Date: Wed, 1 Aug 2018 00:25:20 -0700
+Subject: [PATCH] Add try with resource support
+
+
+diff --git a/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java b/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java
+index 8ac1d92..7212a0a 100644
+--- a/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java
++++ b/src/org/jetbrains/java/decompiler/main/rels/MethodProcessorRunnable.java
+@@ -169,6 +169,10 @@ public class MethodProcessorRunnable implements Runnable {
+ 
+       LabelHelper.identifyLabels(root);
+ 
++      if (TryHelper.enhanceTryStats(root)) {
++        continue;
++      }
++
+       if (InlineSingleBlockHelper.inlineSingleBlocks(root)) {
+         continue;
+       }
+diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/TryHelper.java b/src/org/jetbrains/java/decompiler/modules/decompiler/TryHelper.java
+new file mode 100644
+index 0000000..76ae1aa
+--- /dev/null
++++ b/src/org/jetbrains/java/decompiler/modules/decompiler/TryHelper.java
+@@ -0,0 +1,264 @@
++package org.jetbrains.java.decompiler.modules.decompiler;
++
++import org.jetbrains.java.decompiler.main.DecompilerContext;
++import org.jetbrains.java.decompiler.modules.decompiler.exps.AssignmentExprent;
++import org.jetbrains.java.decompiler.modules.decompiler.exps.ExitExprent;
++import org.jetbrains.java.decompiler.modules.decompiler.exps.Exprent;
++import org.jetbrains.java.decompiler.modules.decompiler.exps.InvocationExprent;
++import org.jetbrains.java.decompiler.modules.decompiler.exps.VarExprent;
++import org.jetbrains.java.decompiler.modules.decompiler.stats.BasicBlockStatement;
++import org.jetbrains.java.decompiler.modules.decompiler.stats.CatchAllStatement;
++import org.jetbrains.java.decompiler.modules.decompiler.stats.CatchStatement;
++import org.jetbrains.java.decompiler.modules.decompiler.stats.RootStatement;
++import org.jetbrains.java.decompiler.modules.decompiler.stats.Statement;
++import org.jetbrains.java.decompiler.struct.gen.VarType;
++import java.util.ArrayList;
++import java.util.List;
++
++public class TryHelper
++{
++  private static final String[] CLOSABLES = {
```

Comment:
`CLOSEABLES`

Reply:
Does this need a list or can it work like #4 and check for `AutoCloseable` instances?

Reply:
...yes, it should do what #4 does now that I think about it.

Reply:
Yeah, that was a derp on my part. Only the one instance check is needed. I’ll clean that up when I get a chance.